lua/nvim-tree/diagnostics.lua

```diff
@@ -51,6 +52,45 @@ function M.update()
     end
   end
 
+  return buffer_severity
+end
+
+local function from_coc()
+  if vim.g.coc_service_initialized ~= 1 then
+    return {}
+  end
+
+  local buffer_severity = {}
+  local diagnostics = {}
+
+  for _, diagnostic in ipairs(vim.fn.CocAction("diagnosticList")) do
+    local bufname = diagnostic.file
+    local severity = severity_levels[diagnostic.severity]
+
+    local severity_list = diagnostics[bufname] or {}
+    table.insert(severity_list, severity)
+    diagnostics[bufname] = severity_list
+	end
+
+  for bufname, severty_list in pairs(diagnostics) do
+    if not buffer_severity[bufname] then
+      local severity = math.min(unpack(severty_list))
+      buffer_severity[bufname] = severity
+    end
+  end
+
+  return buffer_severity
+end
+
+function M.update()
+  local buffer_severity
+  local source = vim.g.nvim_tree_lsp_diagnostics_source or 'nvim_lsp'
```

Comment:
couldn't you just detect if the user uses nvim_lsp or coc ? something like checking `vim.fn.exists('g:did_coc_loaded')` so we don't add additional configuration options :)